// gsg-typeix/build/server/status-code.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @since 1.0.0
 * @enum
 * @name Redirect
 * @description
 * Status enums
 */
var Status;
(function (Status) {
    // 1xx Informational
    Status[Status["Continue"] = 100] = "Continue";
    Status[Status["Switching_Protocols"] = 101] = "Switching_Protocols";
    Status[Status["Processing"] = 102] = "Processing";
    // 2xx Success
    Status[Status["OK"] = 200] = "OK";
    Status[Status["Created"] = 201] = "Created";
    Status[Status["Accepted"] = 202] = "Accepted";
    Status[Status["Non_Authoritative_Information"] = 203] = "Non_Authoritative_Information";
    Status[Status["No_Content"] = 204] = "No_Content";
    Status[Status["Reset_Content"] = 205] = "Reset_Content";
    Status[Status["Partial_Content"] = 206] = "Partial_Content";
    Status[Status["Multi_Status"] = 207] = "Multi_Status";
    Status[Status["Already_Reported"] = 208] = "Already_Reported";
    Status[Status["IM_Used"] = 226] = "IM_Used";
    // 3xx Redirection
    Status[Status["Multiple_Choices"] = 300] = "Multiple_Choices";
    Status[Status["Moved_Permanently"] = 301] = "Moved_Permanently";
    Status[Status["Found"] = 302] = "Found";
    Status[Status["See_Other"] = 303] = "See_Other";
    Status[Status["Not_Modified"] = 304] = "Not_Modified";
    Status[Status["Use_Proxy"] = 305] = "Use_Proxy";
    Status[Status["Switch_Proxy"] = 306] = "Switch_Proxy";
    Status[Status["Temporary_Redirect"] = 307] = "Temporary_Redirect";
    Status[Status["Permanent_Redirect"] = 308] = "Permanent_Redirect";
    // 4xx Client ErrorMessage
    Status[Status["Bad_Request"] = 400] = "Bad_Request";
    Status[Status["Unauthorized"] = 401] = "Unauthorized";
    Status[Status["Payment_Required"] = 402] = "Payment_Required";
    Status[Status["Forbidden"] = 403] = "Forbidden";
    Status[Status["Not_Found"] = 404] = "Not_Found";
    Status[Status["Method_Not_Allowed"] = 405] = "Method_Not_Allowed";
    Status[Status["Not_Acceptable"] = 407] = "Not_Acceptable";
    Status[Status["Proxy_Authentication_Required"] = 408] = "Proxy_Authentication_Required";
    Status[Status["Conflict"] = 409] = "Conflict";
    Status[Status["Gone"] = 410] = "Gone";
    Status[Status["Length_Required"] = 411] = "Length_Required";
    Status[Status["Precondition_Failed"] = 412] = "Precondition_Failed";
    Status[Status["Payload_Too_Large"] = 413] = "Payload_Too_Large";
    Status[Status["URI_Too_Long"] = 414] = "URI_Too_Long";
    Status[Status["Unsupported_Media_Type"] = 415] = "Unsupported_Media_Type";
    Status[Status["Range_Not_Satisfiable"] = 416] = "Range_Not_Satisfiable";
    Status[Status["Expectation_Failed"] = 417] = "Expectation_Failed";
    Status[Status["Im_a_teapot"] = 418] = "Im_a_teapot";
    Status[Status["Misdirected_Request"] = 421] = "Misdirected_Request";
    Status[Status["Unprocessable_Entity"] = 422] = "Unprocessable_Entity";
    Status[Status["Locked"] = 423] = "Locked";
    Status[Status["Failed_Dependency"] = 424] = "Failed_Dependency";
    Status[Status["Upgrade_Required"] = 426] = "Upgrade_Required";
    Status[Status["Precondition_Required"] = 428] = "Precondition_Required";
    Status[Status["Too_Many_Requests"] = 429] = "Too_Many_Requests";
    Status[Status["Request_Header_Fields_Too_Large"] = 431] = "Request_Header_Fields_Too_Large";
    Status[Status["Unavailable_For_Legal_Reasons"] = 451] = "Unavailable_For_Legal_Reasons";
    // 5xx Server ErrorMessage
    Status[Status["Internal_Server_Error"] = 500] = "Internal_Server_Error";
    Status[Status["Not_Implemented"] = 501] = "Not_Implemented";
    Status[Status["Bad_Gateway"] = 502] = "Bad_Gateway";
    Status[Status["Service_Unavailable"] = 503] = "Service_Unavailable";
    Status[Status["Gateway_Time_out"] = 504] = "Gateway_Time_out";
    Status[Status["HTTP_Version_Not_Supported"] = 505] = "HTTP_Version_Not_Supported";
    Status[Status["Variant_Also_Negotiates"] = 506] = "Variant_Also_Negotiates";
    Status[Status["Insufficient_Storage"] = 507] = "Insufficient_Storage";
    Status[Status["Loop_Detected"] = 508] = "Loop_Detected";
    Status[Status["Not_Extended"] = 510] = "Not_Extended";
    Status[Status["Network_Authentication_Required"] = 511] = "Network_Authentication_Required";
})(Status = exports.Status || (exports.Status = {}));
